src/ai/ai.cpp

```diff
@@ -12,6 +12,231 @@
 
 namespace ai {
 
+namespace alliance_factors {
+/*
+	Returns `true` if `other` has one of `n`'s cores
+*/
+bool has_our_core(sys::state& state, dcon::nation_id n, dcon::nation_id other) {
+	for(auto prov : state.world.nation_get_province_ownership(other)) {
+		auto fat = fatten(state.world, prov).get_province();
+		for(auto c : fat.get_core_as_province()) {
```

Comment:
`get_core_by_prov_tag_key` does a hash map lookup for existing cores using the province and national identity id combination, which could replace this inner loop. However, it may be even faster to loop over all the cores of n's identity and check if any of those is owned by other.